splice_timestamp_api/src/main/java/com/splicemachine/timestamp/impl/TimestampOracle.java

```diff
@@ -29,108 +29,115 @@
         this.timestampBlockManager=timestampBlockManager;
         this.blockSize = blockSize * TIMESTAMP_INCREMENT;
-		initialize();
-	}
+        initialize();
+    }
 
     /**
      * Read the current state of the block from the timestampBlockManager
      *
      * @throws TimestampIOException
      */
-	private void initialize() throws TimestampIOException {
-			synchronized(this) {
+    private void initialize() throws TimestampIOException {
+            synchronized(this) {
                 _maxReservedTimestamp = timestampBlockManager.initialize();
-				_timestampCounter.set(_maxReservedTimestamp + TIMESTAMP_INCREMENT);
-			}
-			try {
-				registerJMX();
-			} catch (Exception e) {
+                _timestampCounter.updateAndGet(oldTS -> Math.max(oldTS, _maxReservedTimestamp + TIMESTAMP_INCREMENT));
+            }
+            try {
+                registerJMX();
+            } catch (Exception e) {
                 SpliceLogUtils.error(LOG, "Unable to register Timestamp Generator with JMX. Service will function but metrics will not be available.");
-			}
+            }
     }
 
-	public long getNextTimestamp(boolean refresh, boolean increment) throws TimestampIOException {
-		if (refresh) {
-            _maxReservedTimestamp = timestampBlockManager.initialize();
-			_timestampCounter.set(_maxReservedTimestamp);
-		}
-
-		long nextTS = increment ? _timestampCounter.addAndGet(TIMESTAMP_INCREMENT): _timestampCounter.get();
-		long maxTS = _maxReservedTimestamp; // avoid the double volatile read
-		if (nextTS > maxTS) {
-			reserveNextBlock(maxTS);
-		}
-		_numTimestampsCreated.incrementAndGet(); // JMX metric
-		return nextTS;
-	}
-
-    private void reserveNextBlock(long priorMaxReservedTimestamp) throws TimestampIOException {
-        synchronized(this)  {
-            if (_maxReservedTimestamp > priorMaxReservedTimestamp) return; // some other thread got there first
-            long nextMax = _maxReservedTimestamp + blockSize;
+    public long getCurrentTimestamp() {
+        return _timestampCounter.get();
+    }
+
+    public synchronized void bumpTimestamp(long newTimestamp) throws TimestampIOException {
+        if (newTimestamp > _timestampCounter.get()) {
+            timestampBlockManager.reserveNextBlock(newTimestamp);
+            _maxReservedTimestamp = newTimestamp;
+            _timestampCounter.updateAndGet(oldTS -> Math.max(oldTS, newTimestamp));
+        }
+    }
+
+    public long getNextTimestamp() throws TimestampIOException {
+        long nextTS = _timestampCounter.addAndGet(TIMESTAMP_INCREMENT);
+        if (nextTS > _maxReservedTimestamp) {
+            reserveNextBlock(nextTS);
+        }
+        _numTimestampsCreated.incrementAndGet(); // JMX metric
+        return nextTS;
+    }
+
+    private synchronized void reserveNextBlock(long nextTS) throws TimestampIOException {
+        while (nextTS > _maxReservedTimestamp) {
+            long blockCount = (_maxReservedTimestamp - nextTS) / blockSize + 1;
```

Comment:
Change to nextTS  - _maxReservedTimestamp to make it positive? Just in case next Ts is much larger than _maxReservedTimestamp and blockCount becomes negative.

Reply:
Done